File: database/crud.py
import sqlite3

# Connect to the SQLite database (create it if it doesn't exist)
conn = sqlite3.connect('sales_app.db')

# Create a cursor object to execute SQL queries
cur = conn.cursor()

"""USER TABLE"""

# C: Function to insert a user
def create_user(name, email, password, user_type):
    cur.execute("INSERT INTO Users (name, email, password, user_type) VALUES (?, ?, ?, ?)", 
                (name, email, password, user_type))
    conn.commit()

# R: Function to read users with optional filtering
def read_users(name=None, email=None, user_type=None):
    query = "SELECT * FROM Users WHERE 1=1"  # Start with a basic query
    params = []

    if name is not None:
        query += " AND name = ?"
        params.append(name)
    if email is not None:
        query += " AND email = ?"
        params.append(email)
    if user_type is not None:
        query += " AND user_type = ?"
        params.append(user_type)

    cur.execute(query, params)
    return cur.fetchall()

# U: Function to update a user
def update_user(user_id, name=None, email=None, password=None):
    query = "UPDATE Users SET "
    params = []

    if name is not None:
        query += "name = ?, "
        params.append(name)
    if email is not None:
        query += "email = ?, "
        params.append(email)
    if password is not None:
        query += "password = ?, "
        params.append(password)

    query = query.rstrip(", ") + " WHERE id = ?"
    params.append(user_id)

    cur.execute(query, params)
    conn.commit()

# D: Function to delete a user
def delete_user(user_id):
    cur.execute("DELETE FROM Users WHERE id = ?", (user_id,))
    conn.commit()


"""PRODUCTS TABLE"""

# C: Function to insert a product
def create_product(description, quantity, price):
    cur.execute("INSERT INTO Products (description, quantity, price) VALUES (?, ?, ?)", 
                (description, quantity, price))
    conn.commit()

# R: Function to read products with optional filtering
def read_products(description=None, quantity=None, price=None):
    query = "SELECT * FROM Products WHERE 1=1"  # Start with a basic query
    params = []

    if description is not None:
        query += " AND description = ?"
        params.append(description)
    if quantity is not None:
        query += " AND quantity = ?"
        params.append(quantity)
    if price is not None:
        query += " AND price = ?"
        params.append(price)

    cur.execute(query, params)
    return cur.fetchall()

# U: Function to update a product
def update_product(product_id, description=None, quantity=None, price=None):
    query = "UPDATE Products SET "
    params = []

    if description is not None:
        query += "description = ?, "
        params.append(description)
    if quantity is not None:
        query += "quantity = ?, "
        params.append(quantity)
    if price is not None:
        query += "price = ?, "
        params.append(price)

    query = query.rstrip(", ") + " WHERE id = ?"
    params.append(product_id)

    cur.execute(query, params)
    conn.commit()

# D: Function to delete a product
def delete_product(product_id):
    cur.execute("DELETE FROM Products WHERE id = ?", (product_id,))
    conn.commit()


"""SALES TABLE"""

# C: Function to insert a sale
def create_sale(seller_id, total_value, payment_method, installment, number_of_installments=None, 
                installment_value=None, discount=0):
    cur.execute("""
        INSERT INTO Sales (seller_id, sale_date, total_value, payment_method, installment, 
                           number_of_installments, installment_value, discount) 
        VALUES (?, CURRENT_TIMESTAMP, ?, ?, ?, ?, ?, ?)
    """, (seller_id, total_value, payment_method, installment, number_of_installments, 
          installment_value, discount))
    conn.commit()

# R: Function to read sales with optional filtering
def read_sales(seller_id=None, sale_date=None, total_value=None, payment_method=None, 
               installment=None, number_of_installments=None, discount=None):
    query = "SELECT * FROM Sales WHERE 1=1"  # Start with a basic query
    params = []

    if seller_id is not None:
        query += " AND seller_id = ?"
        params.append(seller_id)
    if sale_date is not None:
        query += " AND sale_date = ?"
        params.append(sale_date)
    if total_value is not None:
        query += " AND total_value = ?"
        params.append(total_value)
    if payment_method is not None:
        query += " AND payment_method = ?"
        params.append(payment_method)
    if installment is not None:
        query += " AND installment = ?"
        params.append(installment)
    if number_of_installments is not None:
        query += " AND number_of_installments = ?"
        params.append(number_of_installments)
    if discount is not None:
        query += " AND discount = ?"
        params.append(discount)

    cur.execute(query, params)
    return cur.fetchall()

# U: Function to update a sale
def update_sale(sale_id, seller_id=None, sale_date=None, total_value=None, 
                payment_method=None, installment=None, number_of_installments=None, 
                installment_value=None, discount=None):
    query = "UPDATE Sales SET "
    params = []

    if seller_id is not None:
        query += "seller_id = ?, "
        params.append(seller_id)
    if sale_date is not None:
        query += "sale_date = ?, "
        params.append(sale_date)
    if total_value is not None:
        query += "total_value = ?, "
        params.append(total_value)
    if payment_method is not None:
        query += "payment_method = ?, "
        params.append(payment_method)
    if installment is not None:
        query += "installment = ?, "
        params.append(installment)
    if number_of_installments is not None:
        query += "number_of_installments = ?, "
        params.append(number_of_installments)
    if installment_value is not None:
        query += "installment_value = ?, "
        params.append(installment_value)
    if discount is not None:
        query += "discount = ?, "
        params.append(discount)

    query = query.rstrip(", ") + " WHERE id = ?"
    params.append(sale_id)

    cur.execute(query, params)
    conn.commit()

# D: Function to delete a sale
def delete_sale(sale_id):
    cur.execute("DELETE FROM Sales WHERE id = ?", (sale_id,))
    conn.commit()


"""SALESITEMS TABLE"""

# C: Function to add products to a sale
def create_sale_item(sale_id, product_id, quantity, price):
    cur.execute("INSERT INTO SalesItems (sale_id, product_id, quantity, price) VALUES (?, ?, ?, ?)", 
                (sale_id, product_id, quantity, price))
    conn.commit()

# R: Function to read products in one or multiple sales with optional filtering
def read_products_in_sale(sale_ids, product_id=None, quantity=None, price=None):
    query = '''
    SELECT SalesItems.id, Products.description, SalesItems.quantity, SalesItems.price
    FROM SalesItems
    JOIN Products ON SalesItems.product_id = Products.id
    WHERE SalesItems.sale_id IN ({})  -- Placeholder for sale IDs
    '''.format(','.join('?' for _ in sale_ids))  # Generate placeholders for sale_ids
    params = list(sale_ids)

    if product_id is not None:
        query += " AND SalesItems.product_id = ?"
        params.append(product_id)
    if quantity is not None:
        query += " AND SalesItems.quantity = ?"
        params.append(quantity)
    if price is not None:
        query += " AND SalesItems.price = ?"
        params.append(price)

    cur.execute(query, params)
    return cur.fetchall()

# U: Function to update a product in a sale
def update_sale_item(sale_item_id, quantity, price):
    cur.execute("UPDATE SalesItems SET quantity = ?, price = ? WHERE id = ?", 
                (quantity, price, sale_item_id))
    conn.commit()

# D: Function to delete a product from a sale
def delete_sale_item(sale_item_id):
    cur.execute("DELETE FROM SalesItems WHERE id = ?", (sale_item_id,))
    conn.commit()

conn.close()

File: database/crud.js
const sqlite3 = require('sqlite3').verbose();

// Connect to the SQLite database (create it if it doesn't exist)
const db = new sqlite3.Database('sales_app.db');

// User Table Functions

// C: Function to insert a user
function createUser(name, email, password, userType) {
    db.run("INSERT INTO Users (name, email, password, user_type) VALUES (?, ?, ?, ?)", 
           [name, email, password, userType]);
}

// R: Function to read users with optional filtering
function readUsers(name, email, userType, callback) {
    let query = "SELECT * FROM Users WHERE 1=1";
    let params = [];

    if (name) {
        query += " AND name = ?";
        params.push(name);
    }
    if (email) {
        query += " AND email = ?";
        params.push(email);
    }
    if (userType) {
        query += " AND user_type = ?";
        params.push(userType);
    }

    db.all(query, params, (err, rows) => {
        callback(err, rows);
    });
}

// U: Function to update a user
function updateUser(userId, name, email, password) {
    let query = "UPDATE Users SET ";
    let params = [];

    if (name) {
        query += "name = ?, ";
        params.push(name);
    }
    if (email) {
        query += "email = ?, ";
        params.push(email);
    }
    if (password) {
        query += "password = ?, ";
        params.push(password);
    }

    query = query.slice(0, -2) + " WHERE id = ?";
    params.push(userId);

    db.run(query, params);
}

// D: Function to delete a user
function deleteUser(userId) {
    db.run("DELETE FROM Users WHERE id = ?", userId);
}

// Product Table Functions

// C: Function to insert a product
function createProduct(description, quantity, price) {
    db.run("INSERT INTO Products (description, quantity, price) VALUES (?, ?, ?)", 
           [description, quantity, price]);
}

// R: Function to read products with optional filtering
function readProducts(description, quantity, price, callback) {
    let query = "SELECT * FROM Products WHERE 1=1";
    let params = [];

    if (description) {
        query += " AND description = ?";
        params.push(description);
    }
    if (quantity) {
        query += " AND quantity = ?";
        params.push(quantity);
    }
    if (price) {
        query += " AND price = ?";
        params.push(price);
    }

    db.all(query, params, (err, rows) => {
        callback(err, rows);
    });
}

// U: Function to update a product
function updateProduct(productId, description, quantity, price) {
    let query = "UPDATE Products SET ";
    let params = [];

    if (description) {
        query += "description = ?, ";
        params.push(description);
    }
    if (quantity) {
        query += "quantity = ?, ";
        params.push(quantity);
    }
    if (price) {
        query += "price = ?, ";
        params.push(price);
    }

    query = query.slice(0, -2) + " WHERE id = ?";
    params.push(productId);

    db.run(query, params);
}

// D: Function to delete a product
function deleteProduct(productId) {
    db.run("DELETE FROM Products WHERE id = ?", productId);
}

// Sales Table Functions

// C: Function to insert a sale
function createSale(sellerId, totalValue, paymentMethod, installment, numberOfInstallments, 
                    installmentValue, discount) {
    db.run(`
        INSERT INTO Sales (seller_id, sale_date, total_value, payment_method, installment, 
                           number_of_installments, installment_value, discount) 
        VALUES (?, CURRENT_TIMESTAMP, ?, ?, ?, ?, ?, ?)`, 
        [sellerId, totalValue, paymentMethod, installment, numberOfInstallments, 
         installmentValue, discount]);
}

// R: Function to read sales with optional filtering
function readSales(sellerId, saleDate, totalValue, paymentMethod, 
                   installment, numberOfInstallments, discount, callback) {
    let query = "SELECT * FROM Sales WHERE 1=1";
    let params = [];

    if (sellerId) {
        query += " AND seller_id = ?";
        params.push(sellerId);
    }
    if (saleDate) {
        query += " AND sale_date = ?";
        params.push(saleDate);
    }
    if (totalValue) {
        query += " AND total_value = ?";
        params.push(totalValue);
    }
    if (paymentMethod) {
        query += " AND payment_method = ?";
        params.push(paymentMethod);
    }
    if (installment) {
        query += " AND installment = ?";
        params.push(installment);
    }
    if (numberOfInstallments) {
        query += " AND number_of_installments = ?";
        params.push(numberOfInstallments);
    }
    if (discount) {
        query += " AND discount = ?";
        params.push(discount);
    }

    db.all(query, params, (err, rows) => {
        callback(err, rows);
    });
}

// U: Function to update a sale
function updateSale(saleId, sellerId, saleDate, totalValue, 
                    paymentMethod, installment, numberOfInstallments, 
                    installmentValue, discount) {
    let query = "UPDATE Sales SET ";
    let params = [];

    if (sellerId) {
        query += "seller_id = ?, ";
        params.push(sellerId);
    }
    if (saleDate) {
        query += "sale_date = ?, ";
        params.push(saleDate);
    }
    if (totalValue) {
        query += "total_value = ?, ";
        params.push(totalValue);
    }
    if (paymentMethod) {
        query += "payment_method = ?, ";
        params.push(paymentMethod);
    }
    if (installment) {
        query += "installment = ?, ";
        params.push(installment);
    }
    if (numberOfInstallments) {
        query += "number_of_installments = ?, ";
        params.push(numberOfInstallments);
    }
    if (installmentValue) {
        query += "installment_value = ?, ";
        params.push(installmentValue);
    }
    if (discount) {
        query += "discount = ?, ";
        params.push(discount);
    }

    query = query.slice(0, -2) + " WHERE id = ?";
    params.push(saleId);

    db.run(query, params);
}

// D: Function to delete a sale
function deleteSale(saleId) {
    db.run("DELETE FROM Sales WHERE id = ?", saleId);
}

// SalesItems Table Functions

// C: Function to add products to a sale
function createSaleItem(saleId, productId, quantity, price) {
    db.run("INSERT INTO SalesItems (sale_id, product_id, quantity, price) VALUES (?, ?, ?, ?)", 
           [saleId, productId, quantity, price]);
}

// R: Function to read products in one or multiple sales with optional filtering
function readProductsInSale(saleIds, productId, quantity, price, callback) {
    let query = `
    SELECT SalesItems.id, Products.description, SalesItems.quantity, SalesItems.price
    FROM SalesItems
    JOIN Products ON SalesItems.product_id = Products.id
    WHERE SalesItems.sale_id IN (${saleIds.map(() => '?').join(', ')})`;
    let params = saleIds;

    if (productId) {
        query += " AND SalesItems.product_id = ?";
        params.push(productId);
    }
    if (quantity) {
        query += " AND SalesItems.quantity = ?";
        params.push(quantity);
    }
    if (price) {
        query += " AND SalesItems.price = ?";
        params.push(price);
    }

    db.all(query, params, (err, rows) => {
        callback(err, rows);
    });
}

// U: Function to update a product in a sale
function updateSaleItem(saleItemId, quantity, price) {
    db.run("UPDATE SalesItems SET quantity = ?, price = ? WHERE id = ?", 
           [quantity, price, saleItemId]);
}

// D: Function to delete a product from a sale
function deleteSaleItem(saleItemId) {
    db.run("DELETE FROM SalesItems WHERE id = ?", saleItemId);
}

// Close the database connection when done
// You can call db.close() when you want to close the database
// db.close();

module.exports = {
    createUser,
    readUsers,
    updateUser,
    deleteUser,
    createProduct,
    readProducts,
    updateProduct,
    deleteProduct,
    createSale,
    readSales,
    updateSale,
    deleteSale,
    createSaleItem,
    readProductsInSale,
    updateSaleItem,
    deleteSaleItem,
};
